File: leet_Design_155_minstack.py
class MinStack:

    def __init__(self):
        self.s = []
        self.m = []

    def push(self, val: int) -> None:
        if not self.s or val <= self.m[-1]:
            self.m.append(val)
        self.s.append(val)

    def pop(self) -> None:
        if self.m[-1] == self.s[-1]:
            self.m.pop()
        self.s.pop()

    def top(self) -> int:
        return self.s[-1]

    def getMin(self) -> int:
        return self.m[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: leet_Design_155_minstack.c
class MinStack {
public:
    
    deque<int>  st;
    deque<int>  ms;
    
    MinStack() {}

    void push(int val)
    {
        if (st.empty() || val <= ms.front())
            ms.push_front(val);
        st.push_front(val);
    }

    void pop()
    {
        if (ms.front() == st.front())
            ms.pop_front();
        st.pop_front();
    }

    int top()
    {
        return st.front();
    }

    int getMin()
    {
        return ms.front();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
